rfcs/0001-fobnail-architecture-change.md

```diff
@@ -0,0 +1,143 @@
+4. Platform finalizes configuration by sending certificate generated from
+   previously provided CSR.
+5. Fobnail Token verifies and saves certificate and provided configuration to
+   its internal memory, then locks configuration and sends status to Platform
+   Owner, finalizing provisioning process.
+6. Token can be unlocked only by reset, which wipes out all configuration,
+   provisioned platforms and all keys.
+
+## Local platform provisioning
+
+Platform provisioning doesn't change much, except that the process is initiated
+by platform instead of Fobnail Token. In order to protect against platform
+re-provisioning itself, Fobnail Token must require user consent to provision
+platform, presumably by pressing button for a few seconds. As an additional
+security measure, platform provisioning could be lockable through Fobnail Token
+APIs.
+
+Previously all requests were sent in a specific order, but since Fobnail Token
+will be a server now, and client is free to send requests in any order, the
+server must handle this case. We want to reduce the amount of global state we
+keep, all API should take as parameter everything that is required to complete
+operation. To achieve this we introduce objects IDs, Fobnail Token may cache
+some objects in memory, like EK or AIK. Client may use object IDs to reference
+these objects. Each object is temporary and is kept at most as long as client is
+connected to the server. Server regularly issues CoAP ping requests to check
+client connectivity.
+
+1. Platform starts provisioning by sending it's EK certificate chain.
+2. Fobnail Token verifies certificate chain and sends EK object ID in response.
+3. Platform generates and sends AIK together with EK object ID.
+4. Fobnail Token responds with AIK challenge and AIK object ID.
+5. Platform performs Credential Activation using the provided challenge and
+   sends results to Fobnail Token together with EK object ID and AIK object ID.
+6. Fobnail verifies results of Credential Activation and sends status to
+   the Platform. AIK is marked as trusted at this point. Fobnail Token creates
+   Provisioning Context and sends its ID (PC ID) to Platform.
+7. Platform signs and sends its metadata to Fobnail, together with PC ID,
+   Fobnail Token verifies metadata signature against AIK bound to specified
+   Provisioning Context.
+8. Fobnail Token verifies metadata and returns status. Metadata is bound to the
+   specified Provisioning Context.
+9. Platform sends RIMs.
+10. Fobnail verifies RIMs and returns status. RIMs are bound to the Provisioning
+    Context.
+11. At this point additional calls may be issued to send more data. This feature
+    is left for future provisioning extensions.
+12. Platform finalizes provisioning by sending special request.
+13. Fobnail verifies all data bound to the Provisioning Context (AIK and RIMs
+    currently) saves it to flash and sends final status.
+14. Platform is provisioned now.
```

Comment:
This is excellent. We need diagram for that.

Reply:
I have added diagrams in plantuml.
In Mermaid we can't use subscripts/superscripts, theoretically it's possible https://github.com/mermaid-js/mermaid/issues/1376 but does not work sequence diagrams.
There are problems with rendering some things, also it's not possible to position legend to the right (Mermaid doesn't support legends at all).
Probably I will have to do this in draw.io when diagrams get approved.

Reply:
Diagram looks great.